File: medium/Permutations.cpp
#include <unordered_map>
#include <vector>

using namespace std;

class Solution { // space :O(n) //

  vector<vector<int>> ans;
  unordered_map<int, bool> vistied;

public:
  vector<vector<int>> permute(vector<int> &nums) {
    vector<int> cur;
    helper(nums, cur);
    return ans;
  }

  void helper(vector<int> &nums, vector<int> &curPer) {
    if (curPer.size() == nums.size()) {
      this->ans.push_back(curPer);
      return;
    }
    for (int i = 0; i < nums.size(); i++) {
      if (not vistied[nums[i]]) {
        curPer.push_back(nums[i]);
        vistied[nums[i]] = true;
        helper(nums, curPer);
        curPer.pop_back();
        vistied[nums[i]] = false;
      }
    }
  }
};

File: 70+/medium/Permutations.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []

        def backtrack(start: int, end: int):
            if start == end:
                ans.append(nums[:])
                return
            for i in range(start, end):
                nums[i], nums[start] = nums[start], nums[i]
                backtrack(start + 1, end)
                nums[start], nums[i] = nums[i], nums[start]

        backtrack(0, len(nums))

        return ans
